src/webgpu/api/validation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,745 @@
+      t.testRun({
+        texture: texture,
+        origin: { x: 0, y: 0, z: 0 },
+        size: { width: 0, height: 0, depth: 0 },
+        offset: 0,
+        bytesPerRow: 1024,
+        rowsPerImage: rowsPerImage,
+        dataSize: 1,
+        method: method,
+      });
+    }, !success);
+  });
+
+g.test('bound_on_rows_per_image')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine(poptions('rowsPerImageInBlock', [0, 1, 2]))
+      .combine(poptions('copyHeightInBlock', [0, 1, 2]))
+      .combine(poptions('copyDepth', [1, 3]))
+  )
+  .fn(async t => {
+    const { rowsPerImageInBlock, copyHeightInBlock, copyDepth, method } = t.params;
+
+    const format = 'rgba8unorm';
+    const rowsPerImage = rowsPerImageInBlock * kTextureFormatInfo[format].blockHeight;
+    const copyHeight = copyHeightInBlock * kTextureFormatInfo[format].blockHeight;
+
+    const texture = t.device.createTexture({
+      size: { width: 4, height: 4, depth: 3 },
+      mipLevelCount: 1,
+      format: format,
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    let success = true;
+    if (rowsPerImage != 0 && rowsPerImage < copyHeight) {
+      success = false;
+    }
+    if (copyDepth > 1 && rowsPerImage < copyHeight) {
+      success = false;
+    }
+
+    t.expectValidationError(() => {
+      t.testRun({
+        texture: texture,
+        origin: { x: 0, y: 0, z: 0 },
+        size: { width: 0, height: copyHeight, depth: copyDepth },
+        offset: 0,
+        bytesPerRow: 1024,
+        rowsPerImage: rowsPerImage,
+        dataSize: 1,
+        method: method,
+      });
+    }, !success);
+  });
+
+g.test('bound_on_offset')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine(poptions('offsetInBlock', [0, 1, 2]))
+      .combine(poptions('dataSizeInBlock', [0, 1, 2]))
+  )
+  .fn(async t => {
+    const { offsetInBlock, dataSizeInBlock, method } = t.params;
+
+    const format = 'rgba8unorm';
+    const offset = offsetInBlock * kTextureFormatInfo[format].bytesPerBlock;
+    const dataSize = dataSizeInBlock * kTextureFormatInfo[format].bytesPerBlock;
+
+    const texture = t.device.createTexture({
+      size: { width: 4, height: 4, depth: 1 },
+      mipLevelCount: 1,
+      format: format,
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    let success = offset <= dataSize;
+
+    t.expectValidationError(() => {
+      t.testRun({
+        texture: texture,
+        origin: { x: 0, y: 0, z: 0 },
+        size: { width: 0, height: 0, depth: 0 },
+        offset: offset,
+        bytesPerRow: 512,
+        rowsPerImage: 4,
+        dataSize: dataSize,
+        method: method,
+      });
+    }, !success);
+  });
+
+/*** WriteTexture-specific tests ***/
```

Comment:
These should probably just be in a separate file `writeTexture.spec.ts`, and no need for the `method` param.

Reply:
Done